pkg-manager/core/test/catalogs.ts

```diff
@@ -0,0 +1,396 @@
+import { type Catalogs } from '@pnpm/catalogs.types'
+import { WANTED_LOCKFILE } from '@pnpm/constants'
+import { createPeersDirSuffix } from '@pnpm/dependency-path'
+import { type Lockfile } from '@pnpm/lockfile-types'
+import { type ProjectId, type ProjectManifest } from '@pnpm/types'
+import { preparePackages } from '@pnpm/prepare'
+import { type MutatedProject, mutateModules, type ProjectOptions } from '@pnpm/core'
+import { getLockfileImporterId } from '@pnpm/lockfile-file'
+import { filterPackagesFromDir } from '@pnpm/workspace.filter-packages-from-dir'
+import { arrayOfWorkspacePackagesToMap } from '@pnpm/workspace.find-packages'
+import readYamlFile from 'read-yaml-file'
+import path from 'path'
+import { testDefaults } from './utils'
+
+/**
+ * A utility to make writing catalog tests easier by reducing boilerplate.
+ */
+class CatalogTestsController {
+  private readonly workspaceDir: string
+  private catalogs: Catalogs = {}
+
+  constructor (readonly pkgs: Array<{ location: string, package: ProjectManifest }>) {
+    preparePackages(pkgs)
+    this.workspaceDir = process.cwd()
+  }
+
+  setCatalogs (catalogs: Catalogs) {
+    this.catalogs = catalogs
+  }
+
+  async install (opts?: { frozenLockfile?: boolean, filter?: readonly string[] }) {
+    const { allProjects } = await filterPackagesFromDir(
+      this.workspaceDir,
+      opts?.filter?.map(parentDir => ({ parentDir })) ?? [])
```

Comment:
why do you need the filterPackagesFromDir function? All the data is already available. The tests in core are low level. You already passed all the required objects to preparePackages.

You can see in multipleImporters.ts tests how it is done. I think duplicating code in tests is fine. It is easier to understand. While helpers force to additionally learn how the helpers work.

Reply:
merged accidentally. Still, filterPackagesFromDir is not needed as a new dev dep in core

Reply:
Got it. I'll remove this from core in a PR now.

> why do you need the filterPackagesFromDir function? All the data is already available. The tests in core are low level. You already passed all the required objects to preparePackages.

I wanted something that re-read the manifests off of disk after the calling the `updateProjectManifest` method. I also wanted to avoid other people having to edit this test if something about the `Project` interface changes.

> You can see in multipleImporters.ts tests how it is done. I think duplicating code in tests is fine. It is easier to understand. While helpers force to additionally learn how the helpers work.

I agree with this principle. Sometimes helpers do make code harder to understand.

Reply:
https://github.com/pnpm/pnpm/pull/8244